pkg/core/resources/apis/meshservice/match.go

```diff
@@ -0,0 +1,114 @@
+package meshservice
+
+import (
+	"fmt"
+
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	meshservice_api "github.com/kumahq/kuma/pkg/core/resources/apis/meshservice/api/v1alpha1"
+	"github.com/kumahq/kuma/pkg/util/maps"
+)
+
+func MatchDataplanesWithMeshServices(
+	dpps []*core_mesh.DataplaneResource,
+	meshServices []*meshservice_api.MeshServiceResource,
+	matchOnlyHealthy bool,
+) map[*meshservice_api.MeshServiceResource][]*core_mesh.DataplaneResource {
+	result := map[*meshservice_api.MeshServiceResource][]*core_mesh.DataplaneResource{}
+
+	dppsByNameByTag, dppsByName := indexDpsForMatching(dpps, matchOnlyHealthy)
+
+	for _, ms := range meshServices {
+		switch {
+		case ms.Spec.Selector.DataplaneRef != nil:
+			result[ms] = matchByRef(ms, dppsByName)
+		case ms.Spec.Selector.DataplaneTags != nil:
+			result[ms] = matchByTags(ms, dppsByNameByTag)
+		default:
+			result[ms] = nil
+		}
+	}
+
+	return result
+}
+
+func indexDpsForMatching(
+	dpps []*core_mesh.DataplaneResource,
+	matchOnlyHealthy bool,
+) (
+	map[string]map[string]*core_mesh.DataplaneResource,
+	map[string]*core_mesh.DataplaneResource,
+) {
+	// Map of tag pair in a context of a mesh (ex. app:redis from default mesh) to dpp name to dpp.
+	dppsByNameByTag := map[string]map[string]*core_mesh.DataplaneResource{}
+	dppsByName := map[string]*core_mesh.DataplaneResource{}
+
+	for _, dpp := range dpps {
+		inbounds := dpp.Spec.GetNetworking().GetInbound()
+		if matchOnlyHealthy {
+			inbounds = dpp.Spec.GetNetworking().GetHealthyInbounds()
+		}
+
+		for _, inbound := range inbounds {
+			for tagName, tagValue := range inbound.GetTags() {
+				tag := fmt.Sprintf("%s;%s;%s", dpp.Meta.GetMesh(), tagName, tagValue)
```

Comment:
Should the format be defined on top as a constant?

Reply:
added functions